pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -468,6 +489,103 @@ protected Blob makeBlob(long oid) throws SQLException {
     return makeBlob(getLong(i));
   }
 
+  private @Nullable Struct getStruct(int i, String typeName) throws SQLException {
+    checkClosed();
+    byte[] value = getRawValue(i);
+    if (value == null) {
+      return null;
+    }
+
+    PgStruct struct = castNonNull(connection.getTypeInfo().getPgStruct(typeName));
+    PgStructField[] structFields = struct.getFields();
+
+    int len = structFields.length;
+    Field[] newFields = new Field[len];
+
+    boolean isBinary = isBinary(i);
+
+    Tuple structRow = isBinary ? castNonNull(fromStructBin(value)) : castNonNull(fromStruct(value));
+
+    Object[] attributes = new Object[len];
+    try (CurrentRow.CurrentRowLock lock = currentRow.switchTo(newFields, structRow)) {
+      for (int j = 0; j < len; j++) {
+        PgStructField fields = struct.getFields()[j];
+        newFields[j] = new Field(fields.getName(), fields.getOID());
+        newFields[j].setFormat(isBinary ? Field.BINARY_FORMAT : Field.TEXT_FORMAT);
+        attributes[j] = getObject(j + 1);
+      }
+    }
+
+    return struct.withAttributes(attributes);
+  }
+
+  public static Tuple fromStructBin(byte[] value) {
+    ByteBuffer buffer = ByteBuffer.wrap(value);
+    int numberOfColumns = buffer.getInt();
+
+    byte[][] data = new byte[numberOfColumns][];
+    for (int i = 0; buffer.hasRemaining(); i++) {
+      int oid = buffer.getInt(); // ignore type
+      int len = buffer.getInt();
+      if (len == -1) {
+        data[i] = null;
+        continue;
+      }
+      data[i] = new byte[len];
+      buffer.get(data[i]);
+    }
+    return new Tuple(data);
+  }
+
+  public static Tuple fromStruct(byte[] value) {
+    // a struct string will always be enclosed in parentheses
+    assert value[0] == '(' && value[value.length - 1] == ')';
+
+    ArrayList<byte[]> fieldValues = new ArrayList<>();
+    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+
+    boolean inQuote = false;
+    byte[] escapeBytes = new byte[] {'\\', '"'};
+    byte b = 0;
+
+    for (int i = 1; i < value.length - 1; i++) {
+      b = value[i];
+
+      boolean isEscapeByte = false;
+      for (byte escB: escapeBytes) {
+        if (b == escB) {
```

Comment:
as there are only 2 characters, a simple OR statement will both be easier to read and faster.
could also consider extracting into a static helper method